detection_rules/rule_loader.py

```diff
@@ -4,45 +4,209 @@
+        assert not self.frozen, f"Unable to add rule {rule.name} {rule.id} to a frozen collection"
+        assert rule.id not in self.id_map, \
+            f"Rule ID {rule.id} for {rule.name} collides with rule {self.id_map.get(rule.id).name}"
+
+        if rule.path is not None:
+            rule.path = rule.path.resolve()
+            assert rule.path not in self.file_map, f"Rule file {rule.path} already loaded"
+            self.file_map[rule.path] = rule
+
+        self.id_map[rule.id] = rule
+        self.rules.append(rule)
+
+    def load_dict(self, obj: dict, path: Optional[Path] = None):
+        contents = TOMLRuleContents.from_dict(obj)
+        rule = TOMLRule(path=path, contents=contents)
+        self.add_rule(rule)
+
+        return rule
+
+    def load_files(self, paths: Iterable[Path]):
+        """Load multiple files into the collection."""
+        for path in paths:
+            self.load_file(path)
+
+    def _deserialize_toml(self, path: Path) -> dict:
+        # use pytoml instead of toml because of annoying bugs
+        # https://github.com/uiri/toml/issues/152
+        # might also be worth looking at https://github.com/sdispater/tomlkit
+        with io.open(str(path.resolve()), "r", encoding="utf-8") as f:
+            toml_dict = pytoml.load(f)
+            self._toml_cache[path] = toml_dict
+            return toml_dict
+
+    def _get_paths(self, directory: Path, recursive=True) -> List[Path]:
+        if recursive:
+            paths = glob.glob(str(directory / "**" / "*.toml"), recursive=True)
+        else:
+            paths = glob.glob(str(directory / "*.toml"))
+
+        return [Path(path) for path in paths]
+
+    def load_file(self, path: Path) -> TOMLRule:
         try:
-            return f(*args, **kwargs)
-        finally:
-            load_rules.clear()
+            path = path.resolve()
+
+            # use the default rule loader as a cache.
+            # if it already loaded the rule, then we can just use it from that
+            if self.__default is not None and self is not self.__default and path in self.__default.file_map:
+                rule = self.__default.file_map[path]
+                self.add_rule(rule)
+                return rule
+
+            obj = self._deserialize_toml(path)
+            return self.load_dict(obj, path=path)
+        except Exception:
+            print(f"Error loading rule in {path}")
+            raise
+
+    # noinspection PyShadowingBuiltins
+    def load_directory(self, directory: Path, recursive=True, toml_filter: Optional[Callable[[dict], bool]] = None):
+        paths = self._get_paths(directory, recursive=recursive)
+        if toml_filter is not None:
+            paths = [path for path in paths if toml_filter(self._deserialize_toml(path))]
 
-    return wrapped
+        self.load_files(paths)
 
+    def load_by_id(self, rule_id: definitions.UUIDString,
+                   directory: Path = Path(RULES_DIR), recursive=True) -> Optional[TOMLRule]:
```

Comment:
maybe better to convert `RULES_DIR` to a `path` so that it is pervasive, then:

```suggestion
                   directory: Path = RULES_DIR, recursive=True) -> Optional[TOMLRule]:
```

Reply:
yeah, i'll update the global

Reply:
a20d92c076a8050e07132a7deeb897781726c5d5